src/passes/Monomorphize.cpp

```diff
@@ -214,31 +219,162 @@ struct CallContext {
+    // (moved into the monomorphized function) remain ordered with respect to
+    // each other, but must be moved past all non-moving things after them. For
+    // example, say we want to move B and D in this list (of expressions in
+    // execution order):
+    //
+    //  A, B, C, D, E
+    //
+    // After moving B and D we end up with this:
+    //
+    //  A, C, E  and executing later in the monomorphized function:  B, D
+    //
+    // Then we must be able to move B past C and E, and D past E. It is simplest
+    // to compute this in reverse order, starting from E and going back, and
+    // then each time we want to move something we can check if it can cross
+    // over all the non-moving effects we've seen so far. To compute this, first
+    // list out the post-order of the expressions, and then we'll iterate in
+    // reverse.
+    struct Lister
+      : public PostWalker<Lister, UnifiedExpressionVisitor<Lister>> {
+      std::vector<Expression*> list;
+      void visitExpression(Expression* curr) { list.push_back(curr); }
+    } lister;
+    // As a quick estimate, we need space for at least the operands.
+    lister.list.reserve(operands.size());
+
+    for (auto* operand : info.call->operands) {
+      lister.walk(operand);
+    }
+
+    // Go in reverse post-order as explained earlier, noting what cannot be
+    // moved into the context, and while accumulating the effects that are not
+    // moving.
+    std::unordered_set<Expression*> unMovable;
+    EffectAnalyzer nonMovingEffects(options, wasm);
+    for (auto i = int64_t(lister.list.size()) - 1; i >= 0; i--) {
+      auto* curr = lister.list[i];
+
+      // This may have been marked as unmovable because of the parent. We do
+      // that because if a parent is unmovable then we can't move the children
+      // into the context (if we did, they would execute after the parent, but
+      // it needs their values).
+      auto currUnMovable = unMovable.count(curr) > 0;
+      if (!currUnMovable) {
+        // This might be movable or unmovable. Check both effect interactions
+        // (as described before, we want to move this past unmovable code) and
+        // reasons intrinsic to the expression itself that might prevent moving.
+        ShallowEffectAnalyzer currEffects(options, wasm, curr);
+        if (currEffects.invalidates(nonMovingEffects) ||
+            !canBeMovedIntoContext(curr, currEffects)) {
+          unMovable.insert(curr);
+          currUnMovable = true;
+        }
+      }
+
+      if (currUnMovable) {
+        // Regardless of whether this was marked unmovable because of the
+        // parent, or because we just found it cannot be moved, accumulate the
+        // effects, and also mark its immediate children (so that we do the same
+        // when we get to them).
+        nonMovingEffects.visit(curr);
+        for (auto* child : ChildIterator(curr)) {
+          unMovable.insert(child);
+        }
+      }
+    }
```

Comment:
Do we get much benefit from moving partial argument expressions into the function? It seems that leaving some of the leaf expressions as `local.get`s of arguments will usually inhibit any further optimization of the full argument expression in the callee beyond what could have been done in the caller.

Reply:
At least sometimes we do. In the example in the top comment, we move the `struct.new` into the called function, where it might not escape and therefore be optimized out entirely. We don't know that value, but we do still see the big picture better when we move code into the target.

Another example: if a call sends `(i32.eqz (local.get $anything))` then we can at least move the `i32.eqz`, and then the optimizer can see that that has at most 1 bit set.

In general, there is more opportunity to optimize the more we move, I think, but you might be right that opportunities decrease with unknown values as leaves. When I look into tuning I might add a flag to control this (though I hope not to need many flags).

Reply:
Makes sense, thanks!